File: lib/neko_auth_web/graph/schema/services.ex
defmodule NekoAuthWeb.Schema.Services do
  use Absinthe.Schema.Notation

  alias NekoAuthWeb.Graph.Resolvers.ServiceResolver

  scalar :date_time do
    parse(fn input ->
      case DateTime.from_iso8601(input.value) do
        {:ok, datetime, _} -> {:ok, datetime}
        _ -> :error
      end
    end)

    serialize(fn date ->
      DateTime.to_iso8601(date)
    end)
  end

  input_object :create_service_input do
    field :name, non_null(:string)
    field :description, non_null(:string)
    field :url, non_null(:string)
    field :icon_url, :string
    field :redirect_uris, non_null(list_of(non_null(:string)))
    field :scopes, non_null(list_of(non_null(:string)))
    field :application_type, non_null(:string)
    field :status, non_null(:boolean)
    field :email_restriction_type, non_null(:string)
    field :restricted_emails, non_null(list_of(non_null(:string)))
  end

  object :service do
    field :id, :id
    field :name, :string
    field :description, :string
    field :url, :string
    field :icon_url, :string
    field :created_at, :date_time
    field :updated_at, :date_time
    field :client_secret, :string
    field :redirect_uris, list_of(:string)
    field :scopes, list_of(:string)
    field :application_type, :string
    field :status, :boolean
    field :email_restriction_type, :string
    field :restricted_emails, list_of(:string)
    field :owner_email, :string

    field :plain_client_secret, :string
  end


  object :service_mutations do
    field :create_service, type: :service do
      arg :input, non_null(:create_service_input)
      resolve &ServiceResolver.create_service/3
    end

    field :delete_service, type: :service do
      arg :id, non_null(:id)
      resolve &ServiceResolver.delete_service/3
    end

    field :update_service, type: :service do
      arg :id, non_null(:id)
      arg :name, :string
      arg :description, :string
      arg :redirect_uris, list_of(:string)
      arg :scopes, list_of(:string)
      arg :application_type, :string
      arg :status, :boolean
      arg :email_restriction_type, :string
      arg :restricted_emails, list_of(:string)

      arg :client_secret, :boolean

      resolve &ServiceResolver.update_service/3
    end
  end

  object :service_query do
    field :services, type: list_of(:service) do
      resolve &ServiceResolver.get_services/3
    end
  end

end

File: assets/js/lib/graph/services.ts
import { gql } from "@apollo/client"
import create_client, { runWithTokens } from "../apollo"

export interface OAuthService {
  id: string
  name: string
  clientSecret?: string
  secretGenerated: boolean
  description: string
  redirectUris: string[]
  scopes: string[]
  applicationType: string
  status: boolean
  emailRestrictionType: "none" | "whitelist" | "blacklist"
  restrictedEmails: string[]
  createdAt: string
}

const CREATE_SERVICE_MUTATION = gql`
  mutation CreateService($input: CreateServiceInput!) {
    createService(input: $input) {
      id
      name
      plainClientSecret: plain_client_secret
      ownerEmail
    }
  }
`

export async function createService(input: {
  name: string
  description: string
  redirectUris: string[]
  scopes: string[]
  applicationType: string
  status: boolean
  emailRestrictionType: "none" | "whitelist" | "blacklist"
  restrictedEmails: string[]
}): Promise<{
  id: string
  plainClientSecret: string
}> {
  return await runWithTokens(async () => {
    const processed_input = { ...input, url: "n/a", iconUrl: "n/a" }

    const variables = {
      input: processed_input
    }
    return (
      await create_client().mutate({
        mutation: CREATE_SERVICE_MUTATION,
        variables: variables
      })
    ).data.createService
  })
}

const GET_SERVICES_QUERY = gql`
  {
    services {
      id
      name
      description
      redirectUris
      scopes
      applicationType
      status
      emailRestrictionType
      restrictedEmails
      createdAt
    }
  }
`

export async function getServices(): Promise<OAuthService[]> {
  return await runWithTokens(async () => {
    const res = await create_client().query({
      query: GET_SERVICES_QUERY,
      variables: undefined
    })
    return (res.data.services as OAuthService[]) || []
  })
}

const DELETE_SERVICE_MUTATION = gql`
  mutation DeleteService($id: ID!) {
    deleteService(id: $id) {
      id
    }
  }
`

export async function deleteService(id: string): Promise<{ id: string }> {
  return await runWithTokens(async () => {
    const variables = {
      id
    }
    return (
      await create_client().mutate({
        mutation: DELETE_SERVICE_MUTATION,
        variables: variables
      })
    ).data
  })
}
export async function generateNewServiceSecret(id: string) {
  const query = gql`
    mutation UpdateService($id: ID!) {
      updateService(id: $id, clientSecret: true) {
        id
        plainClientSecret
      }
    }
  `

  return await runWithTokens(async () => {
    const variables = {
      id
    }
    return (
      await create_client().mutate({
        mutation: query,
        variables: variables
      })
    ).data.updateService.plainClientSecret
  })
}
